From f615a827fc0b9214e90aa4c50f18ae6547676344 Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Sun, 28 Jul 2024 00:32:07 +0200
Subject: [PATCH] Add Dogecoin MANDATORY_SCRIPT_VERIFY_FLAGS and
 STANDARD_SCRIPT_VERIFY_FLAGS (#25)

Dogecoin has different verify flags, resulting in different enforcement rules for the mempool.

Also enable all the legacy rules on regtest when -legacyscriptrules is enabled, not just those ones under SCRIPT_VERIFY_LEGACY_RULES.

diff --git a/src/consensus/activation.cpp b/src/consensus/activation.cpp
index 69f5b0f109191d226d24dc930be4e49c16409a6b..bec2a7bbf37d14cacf9a893c47a93cdc3bbdcb7b 100644
--- a/src/consensus/activation.cpp
+++ b/src/consensus/activation.cpp
@@ -131,3 +131,10 @@ bool IsMengerEnabled(const Consensus::Params &params,
 bool IsDigishieldEnabled(const Consensus::Params &params, int32_t nHeight) {
     return nHeight >= params.digishieldHeight;
 }
+
+// Command-line argument "-legacyscriptrules" will make the node enforce the old
+// script rules (see SCRIPT_VERIFY_LEGACY_RULES).
+bool IsLegacyScriptRulesEnabled(const Consensus::Params &params) {
+    return gArgs.GetBoolArg("-legacyscriptrules",
+                            params.enforceLegacyScriptRules);
+}
diff --git a/src/consensus/activation.h b/src/consensus/activation.h
index bb35431c4afdea63ef8798997ad7f87f70074b3e..9f0f8b497c86b23f5e1f0d677ff539c971fe1e48 100644
--- a/src/consensus/activation.h
+++ b/src/consensus/activation.h
@@ -55,4 +55,7 @@ bool IsCowperthwaiteEnabled(const Consensus::Params &params,
 /** Check if Dogecoin Digishield protocol upgrade has activated. */
 bool IsDigishieldEnabled(const Consensus::Params &params, int32_t nHeight);
 
+/** Check if the legacy script rules are active (not based on height) */
+bool IsLegacyScriptRulesEnabled(const Consensus::Params &params);
+
 #endif // BITCOIN_CONSENSUS_ACTIVATION_H
diff --git a/src/policy/policy.h b/src/policy/policy.h
index 780ac19c2773049bc3552fd0b260d90d534d99d6..0397171f90afa8edde824a8fa1df1528eeb8e15c 100644
--- a/src/policy/policy.h
+++ b/src/policy/policy.h
@@ -69,7 +69,7 @@ static constexpr Amount DEFAULT_MIN_RELAY_TX_FEE_PER_KB(1000 * SATOSHI);
 /**
  * When transactions fail script evaluations under standard flags, this flagset
  * influences the decision of whether to drop them or to also ban the originator
- * (see CheckInputScripts).
+ * (see CheckInputScripts). Unused in Dogecoin, kept for tests.
  */
 static constexpr uint32_t MANDATORY_SCRIPT_VERIFY_FLAGS{
     SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_STRICTENC |
@@ -77,6 +77,10 @@ static constexpr uint32_t MANDATORY_SCRIPT_VERIFY_FLAGS{
     SCRIPT_VERIFY_NULLFAIL | SCRIPT_VERIFY_MINIMALDATA |
     SCRIPT_ENABLE_SCHNORR_MULTISIG | SCRIPT_ENFORCE_SIGCHECKS};
 
+/** Dogecoin: Mandatory script flags */
+static constexpr uint32_t MANDATORY_SCRIPT_VERIFY_FLAGS_LEGACY{
+    SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_LEGACY_RULES};
+
 /**
  * Standard script verification flags that standard transactions will comply
  * with. However scripts violating these flags may still be present in valid
@@ -87,6 +91,8 @@ static constexpr uint32_t MANDATORY_SCRIPT_VERIFY_FLAGS{
  * restrictive flag set that applies in the current / next upgrade, since it is
  * used in numerous parts of the codebase that are unable to access the
  * contextual information of which upgrades are currently active.
+ *
+ * Unused in Dogecoin, kept for tests.
  */
 static constexpr uint32_t STANDARD_SCRIPT_VERIFY_FLAGS{
     MANDATORY_SCRIPT_VERIFY_FLAGS | SCRIPT_VERIFY_DERSIG |
@@ -95,12 +101,27 @@ static constexpr uint32_t STANDARD_SCRIPT_VERIFY_FLAGS{
     SCRIPT_VERIFY_CHECKLOCKTIMEVERIFY | SCRIPT_VERIFY_CHECKSEQUENCEVERIFY |
     SCRIPT_DISALLOW_SEGWIT_RECOVERY | SCRIPT_VERIFY_INPUT_SIGCHECKS};
 
+/** Dogecoin: Standard script flags */
+static constexpr uint32_t STANDARD_SCRIPT_VERIFY_FLAGS_LEGACY{
+    MANDATORY_SCRIPT_VERIFY_FLAGS_LEGACY | SCRIPT_VERIFY_DERSIG |
+    SCRIPT_VERIFY_STRICTENC | SCRIPT_VERIFY_MINIMALDATA |
+    SCRIPT_VERIFY_NULLDUMMY | SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_NOPS |
+    SCRIPT_VERIFY_CLEANSTACK | SCRIPT_VERIFY_MINIMALIF |
+    SCRIPT_VERIFY_NULLFAIL | SCRIPT_VERIFY_CHECKLOCKTIMEVERIFY |
+    SCRIPT_VERIFY_CHECKSEQUENCEVERIFY | SCRIPT_VERIFY_LOW_S |
+    SCRIPT_DISALLOW_SEGWIT_RECOVERY};
+
 /**
  * For convenience, standard but not mandatory verify flags.
  */
 static constexpr uint32_t STANDARD_NOT_MANDATORY_VERIFY_FLAGS{
     STANDARD_SCRIPT_VERIFY_FLAGS & ~MANDATORY_SCRIPT_VERIFY_FLAGS};
 
+/** Dogecoin: Standard non-mandatory flags */
+static constexpr uint32_t STANDARD_NOT_MANDATORY_VERIFY_FLAGS_LEGACY{
+    STANDARD_SCRIPT_VERIFY_FLAGS_LEGACY &
+    ~MANDATORY_SCRIPT_VERIFY_FLAGS_LEGACY};
+
 /**
  * Used as the flags parameter to sequence and nLocktime checks in non-consensus
  * code.
diff --git a/src/test/util/setup_common.cpp b/src/test/util/setup_common.cpp
index a3147b01d111c1d4f51761c5e46cb9f841424647..494a5cd29412a9f73ff7d9178981c38af5a4ad02 100644
--- a/src/test/util/setup_common.cpp
+++ b/src/test/util/setup_common.cpp
@@ -104,6 +104,11 @@ BasicTestingSetup::BasicTestingSetup(
     fs::create_directories(m_path_root);
     m_args.ForceSetArg("-datadir", fs::PathToString(m_path_root));
     gArgs.ForceSetArg("-datadir", fs::PathToString(m_path_root));
+
+    // Dogecoin: Disable legacy (Dogecoin) script rules to not break all tests
+    m_args.ForceSetArg("-legacyscriptrules", "0");
+    gArgs.ForceSetArg("-legacyscriptrules", "0");
+
     gArgs.ClearPathCache();
     {
         SetupServerArgs(m_node);
diff --git a/src/validation.cpp b/src/validation.cpp
index 22dfebb65bf98da284a0b1cda57ad02ef5e59ea0..04483ded69bf5d6b27cf399518e5627dfcf0906a 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -235,13 +235,6 @@ IsReplayProtectionEnabled(const Consensus::Params &params,
            activation_time.value_or(params.sandersActivationTime);
 }
 
-// Command-line argument "-legacyscriptrules" will make the node enforce the old
-// script rules (see SCRIPT_VERIFY_LEGACY_RULES).
-static bool IsLegacyScriptRulesEnabled(const Consensus::Params &params) {
-    return gArgs.GetBoolArg("-legacyscriptrules",
-                            params.enforceLegacyScriptRules);
-}
-
 /**
  * Checks to avoid mempool polluting consensus critical paths since cached
  * signature and script validity results will be reused if we validate this
@@ -740,8 +733,13 @@ bool MemPoolAccept::PreChecks(ATMPArgs &args, Workspace &ws) {
     unsigned int nSize = tx.GetTotalSize();
 
     // Validate input scripts against standard script flags.
-    const uint32_t scriptVerifyFlags =
-        ws.m_next_block_script_verify_flags | STANDARD_SCRIPT_VERIFY_FLAGS;
+    uint32_t scriptVerifyFlags = ws.m_next_block_script_verify_flags;
+    if (IsLegacyScriptRulesEnabled(
+            args.m_config.GetChainParams().GetConsensus())) {
+        scriptVerifyFlags |= STANDARD_SCRIPT_VERIFY_FLAGS_LEGACY;
+    } else {
+        scriptVerifyFlags |= STANDARD_SCRIPT_VERIFY_FLAGS;
+    }
     ws.m_precomputed_txdata = PrecomputedTransactionData{tx};
     if (!CheckInputScripts(tx, state, m_view, scriptVerifyFlags, true, false,
                            ws.m_precomputed_txdata, GetValidationCache(),
@@ -1895,8 +1893,12 @@ bool CheckInputScripts(const CTransaction &tx, TxValidationState &state,
             // This differs from MANDATORY_SCRIPT_VERIFY_FLAGS as it contains
             // additional upgrade flags (see AcceptToMemoryPoolWorker variable
             // extraFlags).
-            uint32_t mandatoryFlags =
-                flags & ~STANDARD_NOT_MANDATORY_VERIFY_FLAGS;
+            uint32_t mandatoryFlags = flags;
+            if (flags & SCRIPT_VERIFY_LEGACY_RULES) {
+                mandatoryFlags &= ~STANDARD_NOT_MANDATORY_VERIFY_FLAGS_LEGACY;
+            } else {
+                mandatoryFlags &= ~STANDARD_NOT_MANDATORY_VERIFY_FLAGS;
+            }
             if (flags != mandatoryFlags) {
                 // Check whether the failure was caused by a non-mandatory
                 // script verification check. If so, ensure we return
@@ -2101,6 +2103,13 @@ static uint32_t GetNextBlockScriptFlags(const CBlockIndex *pindex,
         flags |= SCRIPT_VERIFY_CHECKSEQUENCEVERIFY;
     }
 
+    // If we are on a legacy network (i.e. Dogecoin), keep the old Script rules,
+    // and don't add any BCH/XEC script flags.
+    if (IsLegacyScriptRulesEnabled(consensusparams)) {
+        flags |= SCRIPT_VERIFY_LEGACY_RULES;
+        return flags;
+    }
+
     // If the UAHF is enabled, we start accepting replay protected txns
     if (IsUAHFenabled(consensusparams, pindex)) {
         flags |= SCRIPT_VERIFY_STRICTENC;
@@ -2142,11 +2151,6 @@ static uint32_t GetNextBlockScriptFlags(const CBlockIndex *pindex,
         flags |= SCRIPT_ENABLE_REPLAY_PROTECTION;
     }
 
-    // If we are on a legacy network (i.e. Dogecoin), keep the old Script rules
-    if (IsLegacyScriptRulesEnabled(consensusparams)) {
-        flags |= SCRIPT_VERIFY_LEGACY_RULES;
-    }
-
     return flags;
 }
 
diff --git a/test/functional/feature_legacy_script_flags.py b/test/functional/feature_legacy_script_flags.py
index 00771549c4f389a4efba6eec24eadfc7988bda69..3c07db952cc0b1c08819a927e3e84ae1b9b6eacf 100644
--- a/test/functional/feature_legacy_script_flags.py
+++ b/test/functional/feature_legacy_script_flags.py
@@ -9,12 +9,10 @@ from test_framework.address import (
     SCRIPTSIG_OP_TRUE,
 )
 from test_framework.blocktools import (
-    VERSION_CHAIN_ID_BITS,
     create_block,
     create_coinbase,
     make_conform_to_ctor,
 )
-from test_framework.hash import hash160
 from test_framework.key import ECKey
 from test_framework.messages import COutPoint, CTransaction, CTxIn, CTxOut
 from test_framework.p2p import P2PDataStore
@@ -29,8 +27,10 @@ from test_framework.script import (
     OP_RETURN,
     OP_TRUE,
     CScript,
+    SignatureHash,
+    SignatureHashForkId,
+    hash160,
 )
-from test_framework.signature_hash import SignatureHash, SignatureHashForkId
 from test_framework.test_framework import BitcoinTestFramework
 from test_framework.txtools import pad_tx
 from test_framework.util import assert_raises_rpc_error
@@ -62,7 +62,7 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
                 int(node.getbestblockhash(), 16),
                 create_coinbase(node.getblockcount() + 1),
             )
-            block.nVersion = VERSION_CHAIN_ID_BITS | 4
+            block.nVersion = 4
             block.vtx += txs
             make_conform_to_ctor(block)
             block.hashMerkleRoot = block.calc_merkle_root()
diff --git a/test/functional/feature_legacy_script_rules.py b/test/functional/feature_legacy_script_rules.py
index 5a71868f5d1c00c303ea915da21dc53cfac772f6..e4b022a59876954e40017596e06834af7bde5571 100644
--- a/test/functional/feature_legacy_script_rules.py
+++ b/test/functional/feature_legacy_script_rules.py
@@ -34,12 +34,13 @@ from test_framework.script import (
     OP_SPLIT,
     OP_XOR,
     CScript,
+    SignatureHash,
     SignatureHashForkId,
     hash160,
 )
 from test_framework.test_framework import BitcoinTestFramework
 from test_framework.txtools import pad_tx
-from test_framework.util import assert_equal, assert_raises_rpc_error
+from test_framework.util import assert_raises_rpc_error
 
 
 class LegacyScriptRulesTest(BitcoinTestFramework):
@@ -223,8 +224,8 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
         spend_tx.vin = [CTxIn(COutPoint(int(txid, 16), checksig_out_idx))]
         pad_tx(spend_tx)
         script = CScript([OP_CHECKSIG])
-        sighash = SignatureHashForkId(script, spend_tx, 0, 0x41, 2000)
-        txsig = private_key.sign_schnorr(sighash) + b"\x41"
+        (sighash, _) = SignatureHash(script, spend_tx, 0, 0x01)
+        txsig = private_key.sign_schnorr(sighash) + b"\x01"
         spend_tx.vin[0].scriptSig = CScript([txsig, public_key, script])
 
         # Regtest has SCRIPT_VERIFY_NULLFAIL, which comes in handy here
@@ -237,6 +238,10 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
         peer = check_mined_tx_fails(
             node, peer, spend_tx, "blk-bad-inputs, parallel script check failed"
         )
+        # Schnorr sigs allowed on XEC with FORKID
+        sighash = SignatureHashForkId(script, spend_tx, 0, 0x41, 2000)
+        txsig = private_key.sign_schnorr(sighash) + b"\x41"
+        spend_tx.vin[0].scriptSig = CScript([txsig, public_key, script])
         nonlegacy_node.sendrawtransaction(spend_tx.serialize().hex())
         check_mined_tx_success(nonlegacy_node, nonlegacy_peer, spend_tx)
 
@@ -245,8 +250,8 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
         spend_tx.vin = [CTxIn(COutPoint(int(txid, 16), checksig_out_idx + 1))]
         pad_tx(spend_tx)
         script = CScript([OP_CHECKSIGVERIFY])
-        sighash = SignatureHashForkId(script, spend_tx, 0, 0x41, 2000)
-        txsig = private_key.sign_schnorr(sighash) + b"\x41"
+        (sighash, _) = SignatureHash(script, spend_tx, 0, 0x01)
+        txsig = private_key.sign_schnorr(sighash) + b"\x01"
         spend_tx.vin[0].scriptSig = CScript([True, txsig, public_key, script])
 
         assert_raises_rpc_error(
@@ -258,6 +263,9 @@ class LegacyScriptRulesTest(BitcoinTestFramework):
         peer = check_mined_tx_fails(
             node, peer, spend_tx, "blk-bad-inputs, parallel script check failed"
         )
+        sighash = SignatureHashForkId(script, spend_tx, 0, 0x41, 2000)
+        txsig = private_key.sign_schnorr(sighash) + b"\x41"
+        spend_tx.vin[0].scriptSig = CScript([True, txsig, public_key, script])
         nonlegacy_node.sendrawtransaction(spend_tx.serialize().hex())
         check_mined_tx_success(nonlegacy_node, nonlegacy_peer, spend_tx)
 
